### PR TITLE
Improve Ctrl-C interruption handling

### PR DESCRIPTION
**Pull request by CorentinNeovision**
In case of a keyboard interruption during a non parallel optimization,
will set the proper FAIL log and status for the interrupted trial.

## Motivation

As mentioned in #3336

## Description of the changes

Add an except for `KeyboardInterrupt` during optimization. Also a minor tweaks to a type hint for `func_err`, which is now not always an `Exception` as `KeyboardInterrupt` is only a `BaseException`.

### CODE REVIEW

File: optuna/study/_optimize.py
Comment (himkt) [inline optuna/study/_optimize.py:196]: I personally prefer `Union[Exception,KeyboardInterrupt]` over `BaseException`, because `BaseException` is little bit too general.

**Review by himkt — COMMENTED** [review]
> I think the documentation could be improved, I tried to add a little more info but I couldn't find how to detail the status of interrupted Trials without being overly specific. If there is somewhere better in the doc than the n_trials and timeout doc to explain what kind of interruption leads to what status for the Trial, for sequential or parallel optimization, I'd be happy to do it.

Thank you for the consideration!
It's the just one idea but how about mentioning it in the description section?

```python
    def optimize(
        self,
        ...
        show_progress_bar: bool = False,
    ) -> None:
        """Optimize an objective function.

        Optimization is done by choosing a suitable set of hyperparameter values from a given
        ...
        See also :class:`~optuna.samplers.TPESampler` for more details on 'TPE'.
        +
        + An optimization will be stopped when receiving a termination signal such as SIGINT and SIGTERM.
        + Unlike other signals, trial is automatically failed when receiving SIGINT (ctrl+c).

        Example:
```

**Comment by himkt** [comment]
I'd like @sile to be reviewer as well. If you are not available, please tell me and remove the assignment.

**Comment by sile** [comment]
👍 
I think that I can review this PR within a week.

File: optuna/study/_optimize.py
Comment (sile) [inline optuna/study/_optimize.py:218]: I'm not sure that we should do to that but it might be better to set the result of `sys.exc_info()` to `func_err_fail exc_info` here as the following `Exception` handling branch does.
What do you think of that? > @CorentinNeovision @himkt

File: optuna/study/study.py
Comment (sile) [inline optuna/study/study.py:332]: The current phrase "it also kills the current trial" sounds a bit ambigious to me (some users would not understand what the word "kill" means exactly). 
How about rewriting this sentence or adding a phrase such as "i.e., set the state of the trail to TrialState.FAILED"?

Comment (sile) [reply optuna/study/study.py:332]: Besides, I think that this PR can handle Ctrl+C (SIGINT) but cannot SIGTERM. So, the current doc seems a bit misleading in this aspect. This is just an idea, but we can replace "in which case" with "if the signal is SIGINT (Ctrl+C)" to make the condition clear.

File: optuna/study/_optimize.py
Comment (himkt) [reply optuna/study/_optimize.py:218]: Thank you for the catch. It makes sense. `except` can takes multiple exceptions so we can update here from

```python
    except KeyboardInterrupt as e:
        state = TrialState.FAIL
        func_err = e
    except Exception as e:
        state = TrialState.FAIL
        func_err = e
        func_err_fail_exc_info = sys.exc_info()
```

to

```python
    except (Exception, KeyboardInterrupt) as e:
        state = TrialState.FAIL
        func_err = e
        func_err_fail_exc_info = sys.exc_info()
```

:link: https://docs.python.org/3/tutorial/errors.html#handling-exceptions

File: optuna/study/study.py
Comment (CorentinNeovision) [reply optuna/study/study.py:332]: SIGTERM could probably also be handled, but it would require more work with signal handling, and furthermore probably wouldn't be compatible with windows. So that was not in the scope of my PR

**Comment by sile** [comment]
> maybe it should also be specified that this only works when n_jobs=1 ?

+1

I agree that supportingparallel optimization is too hard to address within this PR. So adding a note document sounds sufficient and nice.

**Comment by himkt** [comment]
> it doesn't change anything in case of parallel optimization, if n_jobs > 1

I thought the change will only affect a single process execution as you said in the original issue. It looks solid to me, although it would be better to document the behavior.

**Comment by himkt** [comment]
Wow, sorry for the duplicated comment. I agree with @sile.

File: optuna/study/study.py
Comment (himkt) [inline optuna/study/study.py:310]: https://github.com/optuna/optuna/runs/5562609091?check_suite_focus=true
This line is too long. Please fix and check if CI pass. :pray:

**Review by sile — APPROVED** [review]
LGTM 👍

**Comment by sile** [comment]
To merge this PR, please wait for another review by @himkt

File: optuna/study/study.py
Comment (himkt) [inline optuna/study/study.py:311]:
```suggestion
        Optimization will be stopped when receiving a termination signal such as SIGINT and
        SIGTERM. Unlike other signals, a trial is automatically and cleanly failed when receiving
```

**Review by himkt — COMMENTED** [review]
@CorentinNeovision I'm sorry to miss your commits after the review and thank you @sile for your kindly ping. Your PR looks good except for minor suggestions. After my suggestions are applied (or not applied with explanations), I'll approve and merge.

File: optuna/study/study.py
Comment (CorentinNeovision) [reply optuna/study/study.py:311]: I'm ok with your suggestions

**Review by himkt — COMMENTED** [review]
LGTM! Thank you @CorentinNeovision for your contribution!

**Pull request merged**
